### PR TITLE
feat: impact metrics histogram

### PR DESCRIPTION
**Pull request by kwasniew**
## About the changes

Adding last metric type histogram. All micro decisions explained inline. The actual change is 250 LOC, the rest is tests.

Also added a new test for all corner case values across all metric types (parametrized test)

### CODE REVIEW

File: internal/impactmetrics/metric_types.go [hunk 1]
```diff
@@ -23,13 +23,27 @@ type GaugeMetricSample struct {
 	Value  float64      `json:"value"`
 }
 
+type BucketEntry struct {
+	Le    interface{} `json:"le"`
```
Comment (kwasniew) [inline internal/impactmetrics/metric_types.go]: can be "+Inf" or float. Interface to avoid writing custom marshalling

File: internal/impactmetrics/metric_types.go [hunk 2]
```diff
@@ -211,16 +225,192 @@ func (g *gaugeImpl) collect() CollectedMetric {
...
+func uniqueSorted(buckets []float64) []float64 {
+	var filtered []float64
+	for _, b := range buckets {
+		if b != math.Inf(1) {
+			filtered = append(filtered, b)
+		}
+	}
+	sort.Float64s(filtered)
+
+	// Deduplicate adjacent elements
```
Comment (kwasniew) [inline internal/impactmetrics/metric_types.go:279]: I decided not to use slices.Compact which is 1.21+ only. My code will work on older versions too

File: internal/impactmetrics/metric_types.go [hunk 3]
```diff
@@ -211,16 +225,192 @@ func (g *gaugeImpl) collect() CollectedMetric {
...
+
+	return &histogramImpl{
+		name:    name,
+		help:    help,
+		buckets: sortedBuckets,
+		values:  map[string]*histogramData{},
+	}
+}
+
+func uniqueSorted(buckets []float64) []float64 {
```
Comment (kwasniew) [inline internal/impactmetrics/metric_types.go:270]: algorithm here is: filter out inf, sort, deduplicate adjacent. I found it's most idiomatic in go

File: internal/impactmetrics/metric_types.go [hunk 4]
```diff
@@ -211,23 +225,197 @@ func (g *gaugeImpl) collect() CollectedMetric {
...
+		counters:   map[string]*counterImpl{},
+		gauges:     map[string]*gaugeImpl{},
+		histograms: map[string]*histogramImpl{},
 	}
 }
 
 func (r *InMemoryMetricRegistry) Counter(name, help string) Counter {
 	r.mu.Lock()
 	defer r.mu.Unlock()
-	if c, exists := r.counters[name]; exists {
```
Comment (kwasniew) [inline internal/impactmetrics/metric_types.go:230]: ok should be more idiomatic

File: internal/impactmetrics/metric_types.go [hunk 5]
```diff
@@ -211,23 +225,202 @@ func (g *gaugeImpl) collect() CollectedMetric {
 	}
 }
 
+type Histogram interface {
+	Observe(value float64, labels MetricLabels)
+	Restore(sample HistogramMetricSample)
```
Comment (kwasniew) [inline internal/impactmetrics/metric_types.go:238]: pragmatic decision similar to node sdk. it makes restore in InMemoryMetricRegistry much simpler. histogram restoration is way more complex than counter and gauge

File: internal/impactmetrics/metric_types.go [hunk 6]
```diff
@@ -320,6 +547,24 @@ func (r *InMemoryMetricRegistry) Restore(metrics []CollectedMetric) {
...
+					}
+					break
+				}
+			}
+
+			h := r.Histogram(m.Name, m.Help, buckets)
+
+			for _, s := range m.Samples {
+				if sample, ok := s.(HistogramMetricSample); ok {
+					h.Restore(sample)
```
Comment (kwasniew) [inline internal/impactmetrics/metric_types.go:552]: delegate to histogram restore

File: internal/impactmetrics/metric_types_test.go [hunk 7]
```diff
@@ -192,6 +193,181 @@ func TestRestoreReinsertsCollectedMetricsIntoTheRegistry(t *testing.T) {
...
+	tests := []struct {
+		name  string
+		value float64
+	}{
+		{"Infinity", math.Inf(1)},
+		{"-Infinity", math.Inf(-1)},
+		{"NaN", math.NaN()},
+	}
+
+	for _, tc := range tests {
```
Comment (kwasniew) [inline internal/impactmetrics/metric_types_test.go:331]: I found that tc may be idiomatic for test case

File: internal/impactmetrics/metric_types_test.go [hunk 8]
```diff
@@ -192,6 +193,181 @@ func TestRestoreReinsertsCollectedMetricsIntoTheRegistry(t *testing.T) {
...
+		},
+	}, emptyCollect)
+
+	registry.Restore(firstCollect)
+
+	restoredCollect := registry.Collect()
+	assert.Equal(t, firstCollect, restoredCollect)
+}
+
+func TestAllMetricOperationsSilentlyDropInvalidValues(t *testing.T) {
```
Comment (kwasniew) [inline internal/impactmetrics/metric_types_test.go:321]: now that we have all metric types we can test dropping extreme values

File: internal/impactmetrics/metric_types.go [hunk 9]
```diff
@@ -211,23 +225,202 @@ func (g *gaugeImpl) collect() CollectedMetric {
...
+func bucketLeToFloat(le interface{}) float64 {
+	switch v := le.(type) {
+	case string:
+		if v == "+Inf" {
+			return math.Inf(1)
+		}
+	case float64:
+		return v
+	}
+	// Data is internally managed (from CollectedMetric), so invalid types should not occur
```
Comment (kwasniew) [inline internal/impactmetrics/metric_types.go]: explaining why no error is thrown

File: internal/impactmetrics/metric_types.go [hunk 10]
```diff
@@ -211,23 +225,202 @@ func (g *gaugeImpl) collect() CollectedMetric {
...
+	mu      sync.Mutex
+	name    string
+	help    string
+	buckets []float64
+	values  map[string]*histogramData
+}
+
+func newHistogram(name, help string, buckets []float64) *histogramImpl {
+	if len(buckets) == 0 {
+		buckets = DefaultHistogramBuckets
```
Comment (sighphyre) [inline internal/impactmetrics/metric_types.go]: Pretty sure this is safe in this incarnation but referencing the slice directly like this is a footgun waiting to happen if a future patch introduces mutations. A copy here is cheap and a nice future proofing layer
```suggestion
	    buckets = append([]float64(nil), DefaultHistogramBuckets...)
```

File: internal/impactmetrics/metric_types.go [hunk 11]
```diff
@@ -211,23 +225,202 @@ func (g *gaugeImpl) collect() CollectedMetric {
...
+}
+
+func (h *histogramImpl) Observe(value float64, labels MetricLabels) {
+	if isInvalidValue(value) {
+		return
+	}
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	key := LabelKey(labels)
```
Comment (sighphyre) [inline internal/impactmetrics/metric_types.go:297]: This is safe to move outside of the mutex scope I think and it's probably often the bulk of the work this function does

File: internal/impactmetrics/metric_types.go (hunk 9 above)
Comment (sighphyre) [reply internal/impactmetrics/metric_types.go]: "This will never happen" cases tend to happen with alarming regularity haha. I don't necessarily have a problem the correctness here but casts make the code harder to reason about and they're generally a smell that suggests a better solution is nearby. Some options, which I think improve the clarity of the data flow:

- The most obvious approach here is just to do this conversion just prior to sending. I think this is probably most correct - we shouldn't be converting things into wire format formats in the heart of the system. I'd really rather we do this honestly. Making this correctly marshallable isn't hard - this is effectively already doing that but on a weird layer with side effects
- Use a proper tagged type rather than an interface here. It's not ideal but at least we keep some type safety

If we're not going to do either of those, it's good to get into the habit of returning errors from functions that are failable - it's super safe in Go because you can't blow out your call stack by accident and you give future peeps the chance to do something about errors without having to rework multiple layers

**Review by sighphyre — APPROVED** [review]
I don't think this is dangerous. The changes I've asked for could be done later so tacking on approve on here and leaving it to judgement or future work

File: internal/impactmetrics/metric_types.go (hunk 2 above)
Comment (sighphyre) [reply internal/impactmetrics/metric_types.go:279]: You don't need to worry about that, we don't support anything below 1.21 anyway. Think this is fine either way though

File: internal/impactmetrics/metric_types.go (hunk 11 above)
Comment (kwasniew) [reply internal/impactmetrics/metric_types.go:297]: good point. will apply this to other function usages for counters and gauges

File: internal/impactmetrics/metric_types.go (hunk 10 above)
Comment (kwasniew) [reply internal/impactmetrics/metric_types.go]: ah of course. excellent point

File: internal/impactmetrics/metric_types.go (hunk 9 above)
Comment (kwasniew) [reply internal/impactmetrics/metric_types.go]: my fear of custom marshaller and unmarshaller led me to suboptimal domain modelling inside this file. Now that I read your comment I think I agree with having clean types in the core and having more boilerplate at the edges.

Comment (kwasniew) [reply internal/impactmetrics/metric_types.go]: I will also improve the typig of Samples []interface{} `json:"samples"` since it suffers from the same problem
